bpf/process/bpf_execve_event.c

```diff
@@ -329,19 +330,29 @@ execve_send(void *ctx)
 		memset(&curr->bin, 0, sizeof(curr->bin));
 #ifdef __LARGE_BPF_PROG
 		// read from proc exe stored at execve time
-		if (event->exe.len <= BINARY_PATH_MAX_LEN) {
+		if (event->exe.len <= BINARY_PATH_MAX_LEN && !event->exe.error) {
 			curr->bin.path_length = probe_read(curr->bin.path, event->exe.len, event->exe.off);
 			if (curr->bin.path_length == 0)
 				curr->bin.path_length = event->exe.len;
+		} else {
+			heap = map_lookup_elem(&execve_heap, &zero);
+			if (heap) {
+				curr->bin.path_length = probe_read_str(curr->bin.path, BINARY_PATH_MAX_LEN, &heap->maxpath);
+				if (curr->bin.path_length > 1) {
+					// don't include the NULL byte in the length
+					curr->bin.path_length--;
+				}
+			}
 		}
 #else
-		// reuse p->args first string that contains the filename, this can't be
-		// above 256 in size (otherwise the complete will be send via data msg)
-		// which is okay because we need the 256 first bytes.
-		curr->bin.path_length = probe_read_str(curr->bin.path, BINARY_PATH_MAX_LEN, &p->args);
-		if (curr->bin.path_length > 1) {
-			// don't include the NULL byte in the length
-			curr->bin.path_length--;
+		// reuse heap->maxpath that contains the filename.
+		heap = map_lookup_elem(&execve_heap, &zero);
+		if (heap) {
+			curr->bin.path_length = probe_read_str(curr->bin.path, BINARY_PATH_MAX_LEN, &heap->maxpath);
+			if (curr->bin.path_length > 1) {
+				// don't include the NULL byte in the length
+				curr->bin.path_length--;
+			}
 		}
```

Comment:
I don't understand the difference here between what you wrote and the existing code. iirc `&p->args` is just the args containing the filename (+ the rest of the args) that was put into `execve_heap` out of implementation details?

Reply:
I answered [here](https://github.com/cilium/tetragon/pull/2718#discussion_r1696680394).